pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -391,173 +409,172 @@ func validateParamArrayIndex(ctx context.Context, params []v1beta1.Param, spec *
 			if p.Default.Type == v1beta1.ParamTypeArray {
 				for _, pattern := range patterns {
 					for i := 0; i < len(p.Default.ArrayVal); i++ {
-						arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
+						arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
 					}
 				}
 			}
 		}
 	}
 
-	// Collect array params lengths from pipeline
+	// Collect array params lengths from params
 	for _, p := range params {
 		if p.Value.Type == v1beta1.ParamTypeArray {
 			for _, pattern := range patterns {
 				for i := 0; i < len(p.Value.ArrayVal); i++ {
-					arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
+					arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
 				}
 			}
 		}
 	}
+	return arrayParamsLengths
+}
 
+// ValidateOutofBoundArrayParams validates if the array indexing params are out of bound
+func ValidateOutofBoundArrayParams(arrayIndexingParams []string, arrayParams map[string]int) error {
 	outofBoundParams := sets.String{}
-
-	// Validate array param in steps fields.
-	validateStepsParamArrayIndexing(spec.Steps, arrayParams, &outofBoundParams)
-
-	// Validate array param in StepTemplate fields.
-	validateStepsTemplateParamArrayIndexing(spec.StepTemplate, arrayParams, &outofBoundParams)
-
-	// Validate array param in build's volumes
-	validateVolumesParamArrayIndexing(spec.Volumes, arrayParams, &outofBoundParams)
-
-	for _, v := range spec.Workspaces {
-		extractParamIndex(v.MountPath, arrayParams, &outofBoundParams)
+	for _, val := range arrayIndexingParams {
+		indexString := substitution.ExtractIndexString(val)
+		idx, _ := substitution.ExtractIndex(indexString)
+		v := substitution.TrimArrayIndex(val)
+		if paramLength, ok := arrayParams[v]; ok {
+			if idx >= paramLength {
+				outofBoundParams.Insert(val)
+			}
+		}
 	}
-
-	validateSidecarsParamArrayIndexing(spec.Sidecars, arrayParams, &outofBoundParams)
-
 	if outofBoundParams.Len() > 0 {
 		return fmt.Errorf("non-existent param references:%v", outofBoundParams.List())
 	}
-
 	return nil
 }
 
-func extractParamIndex(paramReference string, arrayParams map[string]int, outofBoundParams *sets.String) {
+// ExtractArrayIndexingParamRefs extract the array indexing references
+func ExtractArrayIndexingParamRefs(paramReference string, arrayParams map[string]int, arrayIndexingParams *[]string) {
```

Comment:
"arrayParams" is not used in this finction right? as well as other `extractArrayIndexingRefs*()`

Reply:
yes, I missed this one. Thanks!

Reply:
I removed it from all the functions used